File: sample-conference-videochat/sample-conference-videochat/Components/Calls/ToolBar/ToolBar.m
//
//  ToolBar.m
//  sample-conference-videochat
//
//  Created by Injoit on 13/10/15.
//  Copyright © 2015 QuickBlox Team. All rights reserved.
//

#import "ToolBar.h"
#import "CustomButton.h"

@interface ToolBar()

@property (strong, nonatomic) NSMutableArray *buttons;
@property (strong, nonatomic) NSMutableArray *actions;

@end

@implementation ToolBar

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:CGRectZero];
    if (self) {
        
        self.buttons = [NSMutableArray array];
        self.actions = [NSMutableArray array];
        
        [self setBackgroundImage:[[UIImage alloc] init]
              forToolbarPosition:UIToolbarPositionAny
                      barMetrics:UIBarMetricsDefault];
        
        [self setShadowImage:[[UIImage alloc] init]
          forToolbarPosition:UIToolbarPositionAny];
        
        //Default Gray
        self.backgroundColor = UIColor.clearColor;
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    
    self = [super initWithCoder:coder];
    if (self) {
        
        self.buttons = [NSMutableArray array];
        self.actions = [NSMutableArray array];
        
        [self setBackgroundImage:[[UIImage alloc] init]
              forToolbarPosition:UIToolbarPositionAny
                      barMetrics:UIBarMetricsDefault];
        
        [self setShadowImage:[[UIImage alloc] init]
          forToolbarPosition:UIToolbarPositionAny];
        
        //Default Gray
        self.backgroundColor = UIColor.clearColor;
    }
    return self;
}

- (void)removeAllButtons {
    [self.buttons removeAllObjects];
    [self.actions removeAllObjects];
    [self updateItems];
}

- (void)updateItems {
    NSMutableArray *items = [NSMutableArray array];
    UIBarButtonItem *fs =
    [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace
                                                  target:nil
                                                  action:nil];
    [items addObject:fs];

    for (CustomButton *button in self.buttons) {
        UIBarButtonItem *item = [[UIBarButtonItem alloc] initWithCustomView:button];
        [items addObject:item];
        [items addObject:fs];
    }
    [self setItems:items.copy];
}

- (void)addButton:(UIButton *)button action:(void(^)(UIButton *sender))action {
    
    [button addTarget:self
               action:@selector(pressButton:)
     forControlEvents:UIControlEventTouchUpInside];
    
    [self.buttons addObject:button];
    [self.actions addObject:[action copy]];
}

- (void)pressButton:(CustomButton *)button {
    
    NSUInteger idx = [self.buttons indexOfObject:button];
    
    void(^action)(UIButton *sender) = self.actions[idx];
    action(button);
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Toolbar.swift
//
//  Toolbar.swift
//  sample-videochat-webrtc-swift
//
//  Created by Vitaliy Gorbachov on 4/27/18.
//  Copyright © 2018 Quickblox. All rights reserved.
//

import UIKit

class Toolbar: UIToolbar {
    
    var buttons: [UIButton]
    var actions: [(_ sender: UIButton) -> Void]
    
    required init?(coder aDecoder: NSCoder) {
        self.buttons = []
        self.actions = []
        
        super.init(coder: aDecoder)
        
        self.setBackgroundImage(UIImage(), forToolbarPosition: .any, barMetrics: .default)
        self.setShadowImage(UIImage(), forToolbarPosition: .any)
        
        self.backgroundColor = UIColor.white
    }
    
    func updateItems() {
        
        var items: [UIBarButtonItem] = []
        let fs = UIBarButtonItem.init(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
        
        for button in self.buttons {
            
            let item = UIBarButtonItem.init(customView: button)
            items.append(contentsOf: self.items!)
            items.append(fs)
            items.append(item)
        }
        
        items.append(fs)
        self.setItems(items, animated: false)
    }
    
    func addButton(button: UIButton, action:@escaping (_ sender: UIButton) -> Void) {
        
        button.addTarget(self, action: #selector(Toolbar.pressButton(button:)), for: .touchUpInside)
        self.buttons.append(button)
        self.actions.append(action)
    }
    
    @objc func pressButton(button: ToolbarButton) {
        let idx = self.buttons.index(of: button)
        let action = self.actions[idx!]
        action(button)
    }
}
